src/java/main/org/apache/zookeeper/server/quorum/Learner.java

```diff
@@ -254,6 +260,9 @@ protected void connectToLeader(InetSocketAddress addr)
                 }
 
                 sockConnect(sock, addr, Math.min(self.tickTime * self.syncLimit, remainingInitLimitTime));
+                if (self.isSslQuorum())  {
+                    ((SSLSocket) sock).startHandshake();
+                }
                 sock.setTcpNoDelay(nodelay);
                 break;
             } catch (IOException e) {
```

Comment:
Is it possible to get a more typed exception here (like SSLException) when hand shake failed? This serves several purposes: if we know the SSL cert is invalid there is no point retrying; from supportability point of view when hand shake failed, it's possible that what finally gets logged does not reflect the cause of the failure due to how these exceptions are handled (for example we may see in log that initLimit was exceeded while the real culprit is invalid cert).

Reply:
Setting the following system property will give you very detailed SSL logging:
`-Djavax.net.debug=ssl:all`
I'll add it to the documentation.